internal/controller/storageclient_controller.go

```diff
@@ -55,54 +64,130 @@ const (
-	recorder      *utils.EventReporter
-	storageClient *v1alpha1.StorageClient
-
+	Scheme            *runtime.Scheme
 	OperatorNamespace string
+	AvailableCrds     map[string]bool
+
+	log               klog.Logger
+	ctx               context.Context
+	recorder          *utils.EventReporter
+	storageClient     *v1alpha1.StorageClient
+	storageClientHash string
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *StorageClientReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	ctx := context.Background()
-	if err := mgr.GetCache().IndexField(ctx, &v1alpha1.StorageClaim{}, ownerIndexName, func(obj client.Object) []string {
-		refs := obj.GetOwnerReferences()
-		var owners []string
-		for i := range refs {
-			owners = append(owners, string(refs[i].UID))
+	csiDrivers := []string{templates.RBDDriverName, templates.CephFsDriverName}
+	if err := mgr.GetCache().IndexField(ctx, &corev1.PersistentVolume{}, pvClusterIDIndexName, func(o client.Object) []string {
+		pv := o.(*corev1.PersistentVolume)
+		if pv != nil &&
+			pv.Spec.CSI != nil &&
+			slices.Contains(csiDrivers, pv.Spec.CSI.Driver) &&
+			pv.Spec.CSI.VolumeAttributes["clusterID"] != "" {
+			return []string{pv.Spec.CSI.VolumeAttributes["clusterID"]}
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("unable to set up FieldIndexer for PV cluster id: %v", err)
+	}
+	if err := mgr.GetCache().IndexField(ctx, &snapapi.VolumeSnapshotContent{}, vscClusterIDIndexName, func(o client.Object) []string {
+		vsc := o.(*snapapi.VolumeSnapshotContent)
+		if vsc != nil &&
+			slices.Contains(csiDrivers, vsc.Spec.Driver) &&
+			vsc.Status != nil &&
+			vsc.Status.SnapshotHandle != nil {
+			parts := strings.Split(*vsc.Status.SnapshotHandle, "-")
+			if len(parts) == 9 {
+				// second entry in the volumeID is clusterID which is unique across the cluster
+				return []string{parts[2]}
+			}
 		}
-		return owners
+		return nil
 	}); err != nil {
-		return fmt.Errorf("unable to set up FieldIndexer for StorageClaim's owner uid: %v", err)
+		return fmt.Errorf("unable to set up FieldIndexer for VSC csi driver name: %v", err)
+	}
+
+	if r.AvailableCrds[VolumeGroupSnapshotClassCrdName] {
+		if err := mgr.GetCache().IndexField(ctx, &groupsnapapi.VolumeGroupSnapshotContent{}, vgscClusterIDIndexName, func(o client.Object) []string {
+			vgsc := o.(*groupsnapapi.VolumeGroupSnapshotContent)
+			if vgsc != nil &&
+				slices.Contains(csiDrivers, vgsc.Spec.Driver) &&
+				vgsc.Status != nil &&
+				vgsc.Status.VolumeGroupSnapshotHandle != nil {
+				parts := strings.Split(*vgsc.Status.VolumeGroupSnapshotHandle, "-")
+				if len(parts) == 9 {
+					// second entry in the volumeID is clusterID which is unique across the cluster
+					return []string{parts[2]}
+				}
+			}
+			return nil
+		}); err != nil {
+			return fmt.Errorf("unable to set up FieldIndexer for VGSC csi driver name: %v", err)
+		}
 	}
 
+	enqueueStorageClientRequest := handler.EnqueueRequestsFromMapFunc(
```

Comment:
There was a change done to reconcile the storageclaims when the annotation is changed, is this logic moved to storageclient in this PR or its going to be a separate PR?

Reply:
reconcile on annotation change is originally on storageclient and it continue to exists.